### PR TITLE
handle malloc_usable_size()

### PR DESCRIPTION
**Pull request by bratpiorka**
### Types of changes

- [ ] Bugfix (non-breaking change which fixes issue linked in Description above)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)
- [ ] Other

### Checklist

- [ ] Code compiles without errors
- [ ] All tests pass locally with my changes (see TESTING section in CONTRIBUTING file)
- [ ] Created tests which will fail without the change (if possible)
- [ ] Extended the README/documentation (if necessary)
- [ ] All newly added files have proprietary license (if necessary)
- [ ] All newly added files are referenced in MANIFEST files (if necessary)

## Further comments

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/777)

### CODE REVIEW

**Review by PatKamin — CHANGES_REQUESTED** [review]
> Reviewed 16 of 16 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/777)* status: all files reviewed, 4 unresolved discussions (waiting on @bratpiorka, @kilobyte, and @maciekpac)

___
*[src/fast_pool_allocator.c, line 107 at r4](https://reviewable.io/reviews/memkind/memkind/777#-Myw9qlU6bypDmmt1oia:-Myw9qlU6bypDmmt1oib:bk1idjl) ([raw file](https://github.com/memkind/memkind/blob/18d8b107d58f86e3609705e0324f06ed9c12d3ac/src/fast_pool_allocator.c#L107)):*
> ```C
> MEMKIND_EXPORT void fast_pool_allocator_free(FastPoolAllocator *pool, void *ptr)
> {
>     if (ptr == NULL)
> ```

The commit message suggests that this function haven't exited early before this change. However, all I can see here is cosmetics. Either remove this commit or change the commit message.
___
*[src/memkind_memtier.c, line 909 at r2](https://reviewable.io/reviews/memkind/memkind/777#-Myw6HWe4QpajrRvrRuy:-Myw6HWe4QpajrRvrRuz:b-3ekac) ([raw file](https://github.com/memkind/memkind/blob/d00b906519276b38e309346d6fbeb67490a772a3/src/memkind_memtier.c#L909)):*
> ```C
> {
>     size_t size = 0;
>     // TODO: checking policicy this way is a really ugly hack
> ```

nit: policy
___
*[src/mtt_internals.c, line 10 at r2](https://reviewable.io/reviews/memkind/memkind/777#-Myw6yfM5CdK5KOEGb1k:-Myw6yfM5CdK5KOEGb1l:bcybluz) ([raw file](https://github.com/memkind/memkind/blob/d00b906519276b38e309346d6fbeb67490a772a3/src/mtt_internals.c#L10)):*
> ```C
> 
> #include "assert.h"
> #include "string.h"
> ```

What is it needed for?
___
*[src/mtt_internals.c, line 120 at r3](https://reviewable.io/reviews/memkind/memkind/777#-Myw8tm376pQCmJHd0Jv:-Myw8tm376pQCmJHd0Jw:b-ayps7k) ([raw file](https://github.com/memkind/memkind/blob/04ca0ca0dc3d39fd9aeee141e8b10e254936f152/src/mtt_internals.c#L120)):*
> ```C
>     }
> 
>     assert((ptr) && (size > old_size));
> ```

Redundant assert.

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 14 of 16 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 3 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/777)* status: all files reviewed, 3 unresolved discussions (waiting on @kilobyte, @maciekpac, and @PatKamin)

___
*[src/fast_pool_allocator.c, line 107 at r4](https://reviewable.io/reviews/memkind/memkind/777#-Myw9qlU6bypDmmt1oia:-MywHvmyBotfKRv1dIMF:b-m3tpt8) ([raw file](https://github.com/memkind/memkind/blob/18d8b107d58f86e3609705e0324f06ed9c12d3ac/src/fast_pool_allocator.c#L107)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

The commit message suggests that this function haven't exited early before this change. However, all I can see here is cosmetics. Either remove this commit or change the commit message.
</blockquote></details>

you are right - this change was included in my previous PR. I removed this commit
___
*[src/mtt_internals.c, line 10 at r2](https://reviewable.io/reviews/memkind/memkind/777#-Myw6yfM5CdK5KOEGb1k:-MywILkxDqAuoZtQGeNx:be20wdw) ([raw file](https://github.com/memkind/memkind/blob/d00b906519276b38e309346d6fbeb67490a772a3/src/mtt_internals.c#L10)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

What is it needed for?
</blockquote></details>

for memcpy()
___
*[src/mtt_internals.c, line 120 at r3](https://reviewable.io/reviews/memkind/memkind/777#-Myw8tm376pQCmJHd0Jv:-MywJmF952zH9innxIzh:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/04ca0ca0dc3d39fd9aeee141e8b10e254936f152/src/mtt_internals.c#L120)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

Redundant assert.
</blockquote></details>

Done.

File: src/mtt_internals.c [hunk 1]
```diff
@@ -103,8 +104,24 @@ MEMKIND_EXPORT void *mtt_internals_realloc(MttInternals *internals, void *ptr,
...
+    }
+
+    size_t old_size = mtt_internals_usable_size(internals, ptr);
+    if (size <= old_size) {
+        return ptr;
+    }
+
+    void *ret = mtt_internals_malloc(internals, size);
+    memcpy(ret, ptr, old_size);
     mtt_internals_free(internals, ptr);
```
Comment (maciekpac) [inline src/mtt_internals.c:122]: *IMPORTANT* this behavior deviates from  the C standard!

In the case of failed allocation, NULL should be returned but the memory associated with the old pointer should NOT be freed !

[If the function fails to allocate the requested block of memory, a null pointer is returned, and the memory block pointed to by argument ptr is not deallocated (it is still valid, and with its contents unchanged).](https://www.cplusplus.com/reference/cstdlib/realloc/)

Comment (maciekpac) [inline src/mtt_internals.c:122]: *IMPORTANT* the behavior of this function goes against the standard:

[If the function fails to allocate the requested block of memory, a null pointer is returned, and the memory block pointed to by argument ptr is not deallocated (it is still valid, and with its contents unchanged).](https://www.cplusplus.com/reference/cstdlib/realloc/)

File: src/mtt_internals.c [hunk 2]
```diff
@@ -103,8 +104,24 @@ MEMKIND_EXPORT void *mtt_internals_realloc(MttInternals *internals, void *ptr,
...
+
+    if (size == 0) {
+        mtt_internals_free(internals, ptr);
+        return NULL;
+    } else if (ptr == NULL) {
+        return mtt_internals_malloc(internals, size);
+    }
+
+    size_t old_size = mtt_internals_usable_size(internals, ptr);
+    if (size <= old_size) {
```
Comment (maciekpac) [inline src/mtt_internals.c:116]: What if user allocated e.g. 1GB and then reallocates to 1kB ? We would waste a lot of memory.

**Review by maciekpac — CHANGES_REQUESTED** [review]

**Review by PatKamin — APPROVED** [review]
> Reviewed 2 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/777)* status: all files reviewed, 3 unresolved discussions (waiting on @bratpiorka and @kilobyte)

**Review by bratpiorka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/777)* status: 14 of 16 files reviewed, 3 unresolved discussions (waiting on @bratpiorka, @kilobyte, @maciekpac, and @PatKamin)

___
*[src/memkind_memtier.c, line 910 at r4](https://reviewable.io/reviews/memkind/memkind/777#-Myl2tkmf97P0fo3rM3W-r4-910:-Mz0JbFr75DOcB6dtU5n:b-bol8jg) ([raw file](https://github.com/memkind/memkind/blob/18d8b107d58f86e3609705e0324f06ed9c12d3ac/src/memkind_memtier.c#L910)):*
<details><summary><i>Previously, maciekpac wrote…</i></summary><blockquote>

Was it rebased?
</blockquote></details>

yes - but I updated the comment
___
*[src/mtt_internals.c, line 116 at r6](https://reviewable.io/reviews/memkind/memkind/777#-Myl2tkmf97P0fo3rM3X-r6-116:-Mz0Lb5U7jj0L_Y6SgzZ:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/6dc58a00421201c73c9a6674058b1759f6e94a33/src/mtt_internals.c#L116)):*
<details><summary><i>Previously, maciekpac wrote…</i></summary><blockquote>

What if user allocated e.g. 1GB and then reallocates to 1kB ? We would waste a lot of memory.
</blockquote></details>

Done.
___
*[src/mtt_internals.c, line 122 at r6](https://reviewable.io/reviews/memkind/memkind/777#-Myl2tkmf97P0fo3rM3X-r6-122:-Mz0LbVB4th8FUW1fAA6:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/6dc58a00421201c73c9a6674058b1759f6e94a33/src/mtt_internals.c#L122)):*
<details><summary><i>Previously, maciekpac wrote…</i></summary><blockquote>

*IMPORTANT* the behavior of this function goes against the standard:

[If the function fails to allocate the requested block of memory, a null pointer is returned, and the memory block pointed to by argument ptr is not deallocated (it is still valid, and with its contents unchanged).](https://www.cplusplus.com/reference/cstdlib/realloc/)
</blockquote></details>

Done.

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 2 of 2 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/777)* status: all files reviewed, 3 unresolved discussions (waiting on @kilobyte and @maciekpac)

**Review by maciekpac — CHANGES_REQUESTED** [review]
> Reviewed 8 of 16 files at r2, 1 of 3 files at r5, 1 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/777)* status: all files reviewed, 1 unresolved discussion (waiting on @bratpiorka and @kilobyte)

___
*[src/mtt_internals.c, line 122 at r6](https://reviewable.io/reviews/memkind/memkind/777#-Myl2tkmf97P0fo3rM3X-r6-122:-Mz1P6VP1rLNZskKa5tl:bw9o161) ([raw file](https://github.com/memkind/memkind/blob/6dc58a00421201c73c9a6674058b1759f6e94a33/src/mtt_internals.c#L122)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

mtt_internals_realloc just calls free + malloc now, which is not OK..

I think it got reversed to the previous version... Maybe there was an issue with rebase, or am I missing something?

**Review by maciekpac — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/777)* status: all files reviewed, 1 unresolved discussion (waiting on @bratpiorka and @kilobyte)

___
*[src/mtt_internals.c, line 122 at r6](https://reviewable.io/reviews/memkind/memkind/777#-Myl2tkmf97P0fo3rM3X-r6-122:-Mz1QUr75N-houQQ-Nnh:b-c4csf) ([raw file](https://github.com/memkind/memkind/blob/6dc58a00421201c73c9a6674058b1759f6e94a33/src/mtt_internals.c#L122)):*
<details><summary><i>Previously, maciekpac wrote…</i></summary><blockquote>

mtt_internals_realloc just calls free + malloc now, which is not OK..

I think it got reversed to the previous version... Maybe there was an issue with rebase, or am I missing something?
</blockquote></details>

Actually, I think this might not have been a part of this commit... I'm giving approve

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/777)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @kilobyte)

**Pull request merged**